haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs

```diff
@@ -1789,9 +1794,32 @@ instance FromJSON MarkedString where
   parseJSON (A.String t) = pure $ PlainString t
   parseJSON o            = CodeString <$> parseJSON o
 
+-- -------------------------------------
+
+data HoverContents =
+    HoverContentsMS (List MarkedString)
+  | HoverContents   MarkupContent
+  deriving (Read,Show,Eq)
+
+instance ToJSON HoverContents where
+  toJSON (HoverContentsMS  x) = toJSON x
+  toJSON (HoverContents    x) = toJSON x
+instance FromJSON HoverContents where
+  parseJSON v@(A.String _) = HoverContentsMS <$> parseJSON v
+  parseJSON v@(A.Null)     = HoverContentsMS <$> parseJSON v
+  parseJSON v@(A.Array _)  = HoverContentsMS <$> parseJSON v
+  parseJSON v@(A.Object o) = do
+    mk <- o .:? "kind" :: Parser (Maybe MarkupKind)
+    case mk of
+      Nothing -> HoverContentsMS <$> parseJSON v
+      _       -> HoverContents   <$> parseJSON v
+  parseJSON _ = fail "HoverContents"
+
```

Comment:
This `FromJSON` instance is horrible, I welcome any suggestions for improvement.

Reply:
I used `Alternative` when dealing with the `CodeAction`/`Command` overload:

https://github.com/alanz/haskell-lsp/blob/cd94c9bef3b860b6187a3969b0e203a41c0ffef1/haskell-lsp-types/src/Language/Haskell/LSP/Types/CodeAction.hs#L285-L286